### ISSUE TITLE
SVG output

### ISSUE THREAD

**Issue by jbaayen** [Khan/KaTeX#375]
SVG output, as offered by MathJax, would be great to have. Applications include formulas in diagrams, for example.

**Comment by kevinbarabash** [comment]
@jbaayen there's been some work to render to canvas, see https://github.com/Khan/KaTeX/pull/251.  Rendering to other targets is not a huge priority.  We have discussed providing some sort of intermediate representation to make this easier.

**Comment by mbourne** [comment]
@jbaayen: I'm developing SVG graphs, and the approach I'm using when including formulas is to overlay the SVG with HTML elements. There are some examples on this page:

http://www.intmath.com/exponential-logarithmic-functions/calculating-e.php

It suits my purposes.

**Comment by jbaayen** [comment]
@kevinbarabash @mbourne, these sound like steps in the right direction!

We're looking at an out-of-browser application though, where SVG is the graphics format. Given an intermediate representation, we may be able to develop the SVG output ourselves.

**Comment by Locercus** [comment]
I'd just like to add that `foreignObject` is horrendously broken in WebKit and blink. Positioning is completely off.

I.e. see https://stackoverflow.com/a/10353045/1248084

**Comment by pnsaevik** [comment]
If anyone decides to implement SVG output, it would be nice to do it with `<text>` tags instead of outlined glyphs. This way it would be much faster to load. MathJax uses outlined glyphs.

**Comment by gagern** [comment]
@pnsaevik: If one were to use `<text>`, one would have to make sure that the fonts used are the ones shipped with KaTeX. In a browser, it might suffice to use the same CSS file as for KaTeX. For stand-alone SVG, one might have to link to external resources or embed the fonts as `data:` URIs. And one would have to choose between WebFonts and SVG fonts. The latter don't support various features like hinting, but on the other hand SVG fonts are the only font the [SVG 1.1 spec requires](http://www.w3.org/TR/SVG11/fonts.html#Introduction) viewers to support. [Not that it were actually supported in the wild](http://caniuse.com/#feat=svg-fonts), though.

**Comment by pnsaevik** [comment]
@gagern Thanks for your comments! It would definitely have to rely on webfonts, just as regular KaTeX. Are you saying that webfonts within SVG are actually nonstandard? Are there any browsers that support CSS webfonts but *not* SVG webfonts? Sorry that I'm speaking about things where I have only little knowledge..

**Comment by bryevdv** [comment]
> But then, one may argue that CSS is sufficient, and that SVG provides no extra advantage.

I realize this is probably not currently high on any priority list, but just by way of documenting use-cases: Bokeh just added an SVG output option for plots (a *long* requested feature). Users can also render KaTeX labels on plots with an extension, but since these are separate DOM elements, they are missing from the SVG export. We use `canvas2svg` to convert the plot canvas to SVG, and presumably if KaTeX could generate SVG as well, we could find a way to merge things together. This would make users wanting to export Bokeh plots for inclusion in scientific publications very *extremely* happy.

**Comment by jpatel3** [comment]
@mbourne Are you using foreignObject? how are you referring the position and element in SVG?

**Comment by mbourne** [comment]
@jpatel3 No, I'm not using foreignObject. For the particular example I linked to above, I'm using a PHP -> SVG library.

My text creating function has the following parameters:

`function text($pt,$str,$pos,$class,$attrs,$brdParams) {...}`

where $pt is something like [1, 3] and $str is either plain text (which creates an ordinary SVG <text> element), or if it contains HTML tags or KaTeX, it creates an absolute-positioned div that sits on top of the SVG graph at position [1,3], determined by the pixel count from the left and top of the div containing the SVG element, allowing for the position of origin and any padding, using this.

`$htmlTxt = "<div class=\"$class svgHtml\" style=\"top:".round($height - $p[1]*$yunitlength-$oy-$padding)."px;left:".round($p[0]*$xunitlength+$ox)."px\">$str</div>";`

As the whole thing is being created server-side, any KaTeX sitting on top of the SVG is processed along with the rest of the page on page load.

**Comment by fabiospampinato** [comment]
Additional benefits of outputting to SVG: an SVG can base64-encoded and directly piped into an <img> tag, which would lead to fewer DOM nodes on the page (which generally makes things faster) and it also would require essentially 0 sanitization whatsoever, speeding things up that way too.

**Comment by enjoysmath** [comment]
Hello, I would like to tird this proposal.  I also need SVG output.  Right now I'm limited to PNG screen captures of a QWebEnginePage displaying the KaTeX-rendered page.  This is not preferable.  What's preferable is to preserve the zoomability by keeping things SVG, but I don't know how to grab it!

🥇

**Comment by StarHosea** [comment]
really need this feature for ouput pdf

**Comment by mbourne** [comment]
@StarHosea But you can do PDF output of pages containing KaTeX math without SVG. What will SVG give you that HTML doesn't?

**Comment by StarHosea** [comment]
We generate PDF in node server by tools like `pdfmake` to, which doesn't support html .
I am trying to use MathJax to render Tex to svg , and then render to pdf

**Comment by HKGx** [comment]
>I am trying to use MathJax to render Tex to svg , and then render to pdf

@StarHosea, I hacked my way using MathJax to render SVG files in https://github.com/HKGx/rendertex/tree/main/src/services

**Comment by StarHosea** [comment]
@HKGx Thanks a lot , I used `mathjax-node`  like this
```ts
import  mjApi  from 'mathjax-node';
mjApi.start();
const formula = 'k = \frac{2\pi}{\lambda}';
const result = await mjApi.typeset({
    math: formula,
    format: 'TeX',
    svg: true
  });
```

**Comment by enjoysmath** [comment]
How did you hax it, and can you hax it for KaTeX?

**Comment by MohammadRaziei** [comment]
Is there any solutions? I used katex with d3js to create an SVG. but when I export it... none of maths shows in exported svg. so I need to render the input as SVG in foreignObject

**Comment by rambip** [comment]
100% agree.
One use-case would be to use in [manim-web](https://github.com/maloyan/manim-web), the js port of 3b1b manim library.

Getting the SVG path is what makes the smooth and powerful animations possible.

**Comment by varkor** [comment]
In case it is helpful to have more use cases, [**quiver**](https://github.com/varkor/quiver) is a mathematical diagram editor, which uses KaTeX to render its labels. SVG export is the number one requested feature for the editor (https://github.com/varkor/quiver/issues/8), but this is currently blocked on KaTeX supporting SVG as an output format.

**Comment by iamlemec** [comment]
If people are interested, I have a pretty good katex-compatible LaTeX to SVG renderer going with [@gum-jsx/math](https://github.com/CompendiumLabs/gum-jsx-math). It uses actual katex to parse LaTeX into tree form, then renders it to SVG using [gum.jsx](https://github.com/CompendiumLabs/gum-jsx), which is a general purpose visualization language we're developing. You can also interleave math with more complex graphics. There's an live demo at [gum.tex](https://compendiumlabs.ai/gum.tex/).

I'm not sure if you would want to bring in an external dependency like this (it's about 500 kB of JS when built), but perhaps it could provide some inspiration for an in-house katex effort.
